### PR TITLE
[WIP] Flexible CSS transition control

### PR DESCRIPTION
**Pull request by ysds**
This is an experimental PR for #28968.

This allows more control over CSS classes to achieve flexible CSS Transitions. The following GIF shows two transition patterns:

* `fade-xxx` classes: fadeIn/Out
* `slide-down-xxx` classes: slideDown/Up

Also you can use your custom transition name (`my-custom-transition-xxxx)`

transition

Demo: https://codepen.io/anon/pen/dxYqYz?editors=1111

* [x] Dropdown
* [ ] Modal
* [ ] Popover
* [x] Toast
* [x] Tooltip
* [ ] Unit test

### CODE REVIEW

File: js/src/toast.js [hunk 1]
```diff
@@ -96,14 +94,24 @@ class Toast {
       return
     }
 
-    if (this._config.animation) {
-      this._element.classList.add(ClassName.FADE)
+    if (this._config.transitionName) {
+      this._element.classList.add(`${this._config.transitionName}-enter`)
```
Comment (XhmikosR) [inline js/src/toast.js]: You can pass multiple classes IIRC.

Comment (Johann-S) [reply js/src/toast.js]: According to https://caniuse.com/#feat=classlist it seems not 😟

**Comment by Johann-S** [comment]
Nice work! very interesting @ysds 👍

**Comment by ysds** [comment]
Added a demo to the PR's description 😄

**Comment by mdo** [comment]
Big fan of abstracting the transitions. I've long wanted to add more transitions to different elements (grow or scale, slide in different directions, etc).

File: js/src/toast.js [hunk 2]
```diff
@@ -36,22 +37,19 @@ const Event = {
...
-  delay: 'number'
+  delay: 'number',
+  transitionName: 'string'
 }
 
 const Default = {
-  animation: true,
   autohide: true,
-  delay: 500
+  delay: 5000,
```
Comment (MartijnCuppens) [inline js/src/toast.js]: Is this delay increased on purpose?

Comment (ysds) [reply js/src/toast.js]: 😅It's my local test code.

Comment (MartijnCuppens) [reply js/src/toast.js]: “local” 😛

**Comment by MartijnCuppens** [comment]
@ysds, any progress on this?

**Comment by ysds** [comment]
I need help from @twbs/css-review because I'm not expert about JavaScript to apply this change to all components and the related tests correctly. It would be great if anyone could take over this PR.

And what I am wondering is to use a common class (e.g. `fade-enter`) or a component-specific class (e.g. ` modal-enter`) as the default transition class name.

**Comment by MartijnCuppens** [comment]
> And what I am wondering is to use a common class (e.g. `fade-enter`) or a component-specific class (e.g. `modal-enter`) as the default transition class name.

The common classes feel like a better idea, these classes can than be used by multiple components.

**Comment by mdo** [comment]
Closing as stale.

**Pull request closed without merging**